### PR TITLE
ShellClients: Allow X11 clients to use pantheon protocol via mutter hints

### LINKED ISSUE #1965: Launch Dock on X11
### What Happened?

We need a dock for the X11 session since the new dock only works with Wayland

### Steps to Reproduce

Start X11 session and see there's no dock since 956125b488ba3da2ea1b1d93ffa8c0521e14a3ff

### Expected Behavior

Make sure we launch Plank under X11

### OS Version

8.x (Early Access)

### Software Version

Latest release (I have run all updates)

### Log Output

_No response_

### Hardware Info

_No response_

### PR DESCRIPTION
**Pull request by leolost2605**
Fixes #1965 
Closes #1997 

Has the same functionality as #1997 but uses the x property _MUTTER_HINTS

### CODE REVIEW

File: src/ShellClients/ShellClientsManager.vala [hunk 1]
```diff
@@ -36,6 +36,12 @@ public class Gala.ShellClientsManager : Object {
         notifications_client = new NotificationsClient (wm.get_display ());
 
         start_clients.begin ();
+
+        //X11 only
+        wm.get_display ().window_created.connect ((window) => {
+            window.notify["mutter-hints"].connect ((obj, pspec) => parse_mutter_hints ((Meta.Window) obj));
+            parse_mutter_hints (window);
+        });
```
Comment (lenemter) [inline src/ShellClients/ShellClientsManager.vala:44]:
```suggestion
        if (!Meta.Util.is_wayland_compositor ()) {
            wm.get_display ().window_created.connect ((window) => {
                window.notify["mutter-hints"].connect ((obj, pspec) => parse_mutter_hints ((Meta.Window) obj));
                parse_mutter_hints (window);
            });
        }
```

File: src/ShellClients/ShellClientsManager.vala [hunk 2]
```diff
@@ -98,7 +104,15 @@ public class Gala.ShellClientsManager : Object {
...
-    private void make_dock (Meta.Window window) {
+    public void make_dock (Meta.Window window) {
+        if (Meta.Util.is_wayland_compositor ()) {
+            make_dock_wayland (window);
+        } else {
+            make_dock_x11 (window);
+        }
+    }
+
+    private void make_dock_wayland (Meta.Window window) {
```
Comment (lenemter) [inline src/ShellClients/ShellClientsManager.vala:115]:
```suggestion
    private void make_dock_wayland (Meta.Window window) requires (Meta.Util.is_wayland_compositor ()) {
```

File: src/ShellClients/ShellClientsManager.vala [hunk 3]
```diff
@@ -156,4 +189,50 @@ public class Gala.ShellClientsManager : Object {
 
         centered_windows[window] = new CenteredWindow (wm, window);
     }
+
+    //X11 only
+    private void parse_mutter_hints (Meta.Window window) {
```
Comment (lenemter) [inline src/ShellClients/ShellClientsManager.vala:194]:
```suggestion
    private void parse_mutter_hints (Meta.Window window) requires  (!Meta.Util.is_wayland_compositor ()) {
```

**Review by lenemter — COMMENTED** [review]
That's way cleaner then using dbus. Thank you. Still waiting for @tintou review.

File: src/ShellClients/ShellClientsManager.vala [hunk 4]
```diff
@@ -156,4 +190,50 @@ public class Gala.ShellClientsManager : Object {
...
+            if (split.length != 2) {
+                continue;
+            }
+
+            var key = split[0];
+            var val = split[1];
+
+            switch (key) {
+                case "anchor":
+                    set_anchor (window, int.parse (val));
```
Comment (tintou) [inline src/ShellClients/ShellClientsManager.vala:213]: I would use `try_parse` just to be on the safe side

Comment (leolost2605) [reply src/ShellClients/ShellClientsManager.vala]: Done :)

**Review by lenemter — CHANGES_REQUESTED** [review]

**Comment by lenemter** [comment]
Also, I noticed that this PR doesn't allow using Gtk4 notifications server in X11. If we will use this approach with mutter hints, we will have to add a new parameter or something like that to identify the windows. But we can deal with that later.

File: src/ShellClients/ShellClientsManager.vala [hunk 5]
```diff
@@ -156,4 +190,63 @@ public class Gala.ShellClientsManager : Object {
...
+
+            switch (key) {
+                case "anchor":
+                    int parsed;
+                    if (int.try_parse (val, out parsed) && 0 <= parsed && parsed <= 15) {
+                        set_anchor (window, parsed);
+                    } else {
+                        warning ("Failed to parse %s as anchor", val);
+                    }
+                    break;
```
[the remark this answers is not in the document]
Comment (leolost2605) [reply src/ShellClients/ShellClientsManager.vala]: Added a comment :)

set_anchor uses Meta.Side which is a 4 value bitfield so it can go up to 15

**Comment by leolost2605** [comment]
I think regarding the notifications server we can just use the x library to set our window type as notification just like the gtk3 version does

File: src/ShellClients/ShellClientsManager.vala [hunk 6]
```diff
@@ -156,4 +190,63 @@ public class Gala.ShellClientsManager : Object {
...
+
+                case "hide-mode":
+                    int parsed;
+                    if (int.try_parse (val, out parsed) && 0 <= parsed && parsed <= 4) {
+                        set_hide_mode (window, parsed);
+                    } else {
+                        warning ("Failed to parse %s as hide mode", val);
+                    }
+                    break;
```
[the remark this answers is not in the document]
Comment (leolost2605) [reply src/ShellClients/ShellClientsManager.vala]: Done :)

**Review by lenemter — APPROVED** [review]
Code looks good :tada:

**Pull request merged**